src/main/kotlin/com/terraformation/backend/accelerator/db/CohortStore.kt

```diff
@@ -0,0 +1,136 @@
+import org.jooq.Condition
+import org.jooq.DSLContext
+import org.jooq.Field
+import org.jooq.impl.DSL
+import org.springframework.dao.DataIntegrityViolationException
+
+@Named
+class CohortStore(
+    private val clock: InstantSource,
+    private val dslContext: DSLContext,
+    private val cohortsDao: CohortsDao,
+) {
+  fun fetchOneById(
+      cohortId: CohortId,
+      depth: CohortDepth? = CohortDepth.Cohort
+  ): ExistingCohortModel {
+    return fetch(COHORTS.ID.eq(cohortId), depth).firstOrNull()
+        ?: throw CohortNotFoundException(cohortId)
+  }
+
+  fun findAll(depth: CohortDepth? = CohortDepth.Cohort): List<ExistingCohortModel> {
+    return fetch(null, depth)
+  }
+
+  fun create(model: NewCohortModel): ExistingCohortModel {
+    requirePermissions { createCohort() }
+
+    val now = clock.instant()
+    val userId = currentUser().userId
+
+    val row =
+        CohortsRow(
+            createdBy = userId,
+            createdTime = now,
+            modifiedBy = userId,
+            modifiedTime = now,
+            name = model.name,
+            phaseId = model.phase)
+
+    cohortsDao.insert(row)
+
+    return row.toModel()
+  }
+
+  fun delete(cohortId: CohortId) {
+    requirePermissions { deleteCohort(cohortId) }
+
+    try {
+      dslContext.transaction { _ ->
+        val rowsDeleted = dslContext.deleteFrom(COHORTS).where(COHORTS.ID.eq(cohortId)).execute()
+
+        if (rowsDeleted == 0) {
+          throw CohortNotFoundException(cohortId)
+        }
+      }
+    } catch (e: DataIntegrityViolationException) {
+      val hasParticipants =
+          dslContext.fetchExists(PARTICIPANTS, PARTICIPANTS.COHORT_ID.eq(cohortId))
+      if (hasParticipants) {
+        throw CohortHasParticipantsException(cohortId)
+      } else {
+        throw e
+      }
+    }
+  }
+
+  fun update(cohortId: CohortId, updateFunc: (ExistingCohortModel) -> ExistingCohortModel) {
+    requirePermissions { updateCohort(cohortId) }
+
+    val existing = fetchOneById(cohortId)
+    val updated = updateFunc(existing)
+
+    val rowsUpdated =
+        with(COHORTS) {
+          dslContext
+              .update(COHORTS)
+              .set(MODIFIED_BY, currentUser().userId)
+              .set(MODIFIED_TIME, clock.instant())
+              .set(NAME, updated.name)
+              .set(PHASE_ID, updated.phase)
+              .where(ID.eq(cohortId))
+              .execute()
+        }
+
+    if (rowsUpdated < 1) {
+      throw CohortNotFoundException(cohortId)
+    }
+  }
+
+  private fun participantIdsMultiset(): Field<Set<ParticipantId>> =
+      DSL.multiset(
+              DSL.select(PARTICIPANTS.ID)
+                  .from(PARTICIPANTS)
+                  .where(PARTICIPANTS.COHORT_ID.eq(COHORTS.ID))
+                  .orderBy(PARTICIPANTS.ID))
+          .convertFrom { result -> result.map { it[PARTICIPANTS.ID.asNonNullable()] }.toSet() }
+
+  private fun fetch(condition: Condition?, depth: CohortDepth?): List<ExistingCohortModel> {
```

Comment:
Could `depth` be non-nullable here?

Reply:
Yes, it should be. I conflated nullability with default value in the API and plumbed it through as nullable. I switched it back here and in the API to be non-nullable.